pallets/pools/src/lib.rs

```diff
@@ -902,12 +902,71 @@ pub mod pallet {
 
 				pool.start_next_epoch(now)?;
 
+				let mut pool_value = total_assets;
+				// Update the tranches here and note losses and new expected value (i.e. debt)
+				pool.tranches.combine_mut_non_residual_top(|tranche| {
+					if tranche.tranche_type == TrancheType::Residual {
+						tranche.debt = pool_value;
+					} else {
+						tranche.accrue(now)?;
+
+						// Pool value can not satisfy the expected value of the tranche
+						if tranche.expected_assets()? > pool_value {
+							// This arm matches if the pool_value can only partially fill the debt of
+							// a tranche
+							if tranche.debt > pool_value {
+								// Losses increase by what the diff between current value and expected value
+								tranche.loss = tranche
+									.loss
+									.saturating_add(tranche.debt.saturating_sub(pool_value))
+									.saturating_add(tranche.reserve);
+								// We assume the pool will still be able to generate enough value for
+								// satisfying the tranche. This boils down to the assumption of
+								// NAV_{t + 1} >= NAV_{t} * rate_per_sec_{tranche} for every tranche
+								//
+								// Assuming the residual tranche has no rate_per_sec
+								tranche.debt = pool_value;
+								tranche.reserve = Zero::zero();
+								pool_value = Zero::zero();
+
+							// This arm matches if the new pool_value can satisfy the debt and the reserve of tranche
+							// This favors debt over reserve as we want as much value to accrue interest for a tranche
+							} else if tranche.debt.saturating_add(tranche.reserve) > pool_value {
+								let loss = pool_value
+									.saturating_sub(tranche.debt.saturating_add(tranche.reserve));
+								// Losses are partially re-covered
```

Comment:
I think this comment is misleading, in this case there's no recovery of losses, right?

```suggestion
								// Losses are partially increased```